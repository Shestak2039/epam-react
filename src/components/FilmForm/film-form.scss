@import "../../styles/colors";

.film-form {
  padding: 40px;
  width: 500px;

  .header-title {
    color: $default-text-color;
    margin-bottom: 20px;
  }
}

.buttons-area {
  display: flex;
  justify-content: flex-end;
  margin-top: 30px;
}
