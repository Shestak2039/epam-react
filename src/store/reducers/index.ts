import { combineReducers } from 'redux';

import filmsReducer from './films-reducer';

export default combineReducers({
    filmsState: filmsReducer
});
