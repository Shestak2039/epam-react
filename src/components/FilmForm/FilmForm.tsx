import React, { Dispatch } from 'react';

import './film-form.scss';

import { GenreModel } from '../SortingFilms/SortingByGenre/SortingByGenre';

import Button from '../common/Button/Button';
import FillFormInput from './FillFormInput/FillFormInput';
import FillFormSelect from './FillFormSelect/FillFormSelect';
import { connect } from 'react-redux';
import { FilmCardModel } from '../../models/film-card.model';
import { createFilmAsync, updateFilmAsync } from '../../store/actions/film-actions';
import { Formik, Form, FormikValues } from 'formik';
import * as Yup from 'yup';

const mockGenres: GenreModel[] = [
    {id: 1, name: 'ALL', isSelected: true},
    {id: 2, name: 'DOCUMENTARY', isSelected: false},
    {id: 3, name: 'COMEDY', isSelected: false},
    {id: 4, name: 'HORROR', isSelected: false},
    {id: 5, name: 'CRIME', isSelected: false}
];

interface FilmFormProps {
    createFilm: (film: FilmCardModel) => void;
    updateFilm: (filmId: string, film: FilmCardModel) => void;
    closeModal: () => void;
    filmInfo?: FilmCardModel;
    isUpdatingForm?: boolean;
}

const FilmForm: React.FunctionComponent<FilmFormProps> = ({createFilm, updateFilm, closeModal, filmInfo, isUpdatingForm}) => {
    const handleSubmitButton = (values: FormikValues = {}) => {
        const {title, genres, date, movieURL, rating, runtime, overview} = values;
        if (isUpdatingForm) {
            updateFilm(filmInfo?.id || '', {
                title,
                genres: [genres],
                year: date,
                imageUrl: movieURL,
                rating: rating,
                durationInMinutes: runtime,
                overview: overview
            });
        } else {
            createFilm({
                title,
                genres: [genres],
                year: date,
                imageUrl: movieURL,
                rating: rating,
                durationInMinutes: runtime,
                overview: overview
            });
        }
        closeModal();
    };

    return (
        <Formik
            initialValues={{
                title: filmInfo?.title || '',
                date: filmInfo?.year || '',
                movieURL: filmInfo?.imageUrl || '',
                overview: filmInfo?.overview || '',
                genres: filmInfo?.genres[0] || '',
                runtime: filmInfo?.durationInMinutes || 0,
                rating: filmInfo?.rating || 0
            }}
            validationSchema={Yup.object({
                title: Yup.string().min(2, 'Must be 2 characters or more').required(),
                date: Yup.string().matches(/\d\d\d\d-\d\d-\d\d/, 'Must be a date').required(),
                movieURL: Yup.string().matches(/^(?:([A-Za-z]+):)?(\/{0,3})([0-9.\-A-Za-z]+)(?::(\d+))?(?:\/([^?#]*))?(?:\?([^#]*))?(?:#(.*))?$/, 'Must be a link').required(),
                genres: Yup.string().required(),
                overview: Yup.string().required(),
                runtime: Yup.number().required(),
                rating: Yup.number().max(5, 'Max value is 5').min(0).required()
            })}
            onSubmit={(values) => {}}
        >
            {({values, resetForm}) => (
                <Form className="film-form" data-testid="form">
                    <FillFormInput name="title" type="text" placeholder={'Select Title'} label={'TITLE'}/>
                    <FillFormInput name="date" type="date" placeholder={'Select Date'} label={'RELEASE DATE'}/>
                    <FillFormInput name="movieURL" type="text" placeholder={'Movie URL here'} label={'MOVIE URL'}/>
                    <FillFormSelect  label={'GENRE'} name="genres">
                        <option value="">Select a genre</option>
                        <option value="ALL">ALL</option>
                        <option value="DOCUMENTARY">DOCUMENTARY</option>
                        <option value="COMEDY">COMEDY</option>
                        <option value="HORROR">HORROR</option>
                        <option value="CRIME">CRIME</option>
                    </FillFormSelect>
                    <FillFormInput name="overview" type="text" placeholder={'Runtime here'} label={'OVERVIEW'}/>
                    <FillFormInput name="runtime" type="number" placeholder={'Overview here'} label={'RUNTIME'}/>
                    <FillFormInput name="rating" type="number" placeholder={'Rating here'} label={'RATING'}/>
                    <div className="buttons-area">
                        <Button title={'RESET'} primary={false} action={() => resetForm()}/>
                        <Button title={'SUBMIT'} primary={true} type={'submit'} action={() => handleSubmitButton(values)}/>
                    </div>
                </Form>
            )}
        </Formik>
    );
};

const mapDispatchToProps = (dispatch: Dispatch<any>) => {
    return {
        createFilm: (film: FilmCardModel) => dispatch(createFilmAsync(film)),
        updateFilm: (filmId: string, film: FilmCardModel) => dispatch(updateFilmAsync(filmId, film))
    };
};

export default connect(null, mapDispatchToProps)(FilmForm);
