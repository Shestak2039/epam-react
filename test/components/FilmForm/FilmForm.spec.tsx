import React from 'react';
import { render, fireEvent, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import FilmForm from '../../../src/components/FilmForm/FilmForm';
import { Provider } from 'react-redux';
import store from '../../../src/store/store';

describe('FilmForm', () => {
    const closeModal = jest.fn();
    const createFilm = jest.fn();
    const updateFilm = jest.fn();
    
    const renderFillForm = (isUpdatingForm = false) => render(
        <Provider store={store}>
            <FilmForm
                closeModal={closeModal}
                isUpdatingForm={isUpdatingForm}
                // @ts-ignore
                createFilm={createFilm}
                // @ts-ignore
                updateFilm={updateFilm}
            />
        </Provider>
    );

    test('renders FillForm snapshot', () => {
        const {asFragment} = renderFillForm();

        expect(asFragment()).toMatchSnapshot();
    });

    test('should change value in input', () => {
        const { getByPlaceholderText } = renderFillForm();
        
        const titleInput = getByPlaceholderText(/Select Title/i);
        const mockTitle = 'mockTitle';
        
        expect(screen.queryByText(mockTitle)).toBeNull();
        userEvent.type(titleInput, mockTitle);
        expect(titleInput.getAttribute('value')).toEqual(mockTitle);
    });

    it('should send use submit handler and createFilm', async () => {
        const { getByTestId } = renderFillForm();
        const mockSubmitFn = jest.fn();

        const form = getByTestId('form');
        form.onsubmit = mockSubmitFn;

        fireEvent.submit(form);
        userEvent.click(getByTestId('SUBMIT'));

        await waitFor(() => {
            expect(mockSubmitFn).toHaveBeenCalled();
            expect(closeModal).toHaveBeenCalled();
        });
    });
    
    it('should send use submit handler and updateFilm', async () => {
        const { getByTestId } = renderFillForm(true);
        const mockSubmitFn = jest.fn();

        const form = getByTestId('form');
        form.onsubmit = mockSubmitFn;

        fireEvent.submit(form);
        userEvent.click(getByTestId('SUBMIT'));

        await waitFor(() => {
            expect(mockSubmitFn).toHaveBeenCalled();
            expect(closeModal).toHaveBeenCalled();
        });
    });
});
