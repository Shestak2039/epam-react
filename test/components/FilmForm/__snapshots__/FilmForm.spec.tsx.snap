// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmForm renders FillForm snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="film-form"
    data-testid="form"
  >
    <div
      class="form-element"
    >
      <label
        class="label"
        for="title"
      >
        TITLE
      </label>
      <input
        class="input"
        name="title"
        placeholder="Select Title"
        type="text"
        value=""
      />
    </div>
    <div
      class="form-element"
    >
      <label
        class="label"
        for="date"
      >
        RELEASE DATE
      </label>
      <input
        class="input"
        name="date"
        placeholder="Select Date"
        type="date"
        value=""
      />
    </div>
    <div
      class="form-element"
    >
      <label
        class="label"
        for="movieURL"
      >
        MOVIE URL
      </label>
      <input
        class="input"
        name="movieURL"
        placeholder="Movie URL here"
        type="text"
        value=""
      />
    </div>
    <div
      class="form-element"
    >
      <label
        class="label"
        for="genres"
      >
        GENRE
      </label>
      <select
        class="select"
        name="genres"
      >
        <option
          value=""
        >
          Select a genre
        </option>
        <option
          value="ALL"
        >
          ALL
        </option>
        <option
          value="DOCUMENTARY"
        >
          DOCUMENTARY
        </option>
        <option
          value="COMEDY"
        >
          COMEDY
        </option>
        <option
          value="HORROR"
        >
          HORROR
        </option>
        <option
          value="CRIME"
        >
          CRIME
        </option>
      </select>
    </div>
    <div
      class="form-element"
    >
      <label
        class="label"
        for="overview"
      >
        OVERVIEW
      </label>
      <input
        class="input"
        name="overview"
        placeholder="Runtime here"
        type="text"
        value=""
      />
    </div>
    <div
      class="form-element"
    >
      <label
        class="label"
        for="runtime"
      >
        RUNTIME
      </label>
      <input
        class="input"
        name="runtime"
        placeholder="Overview here"
        type="number"
        value="0"
      />
    </div>
    <div
      class="form-element"
    >
      <label
        class="label"
        for="rating"
      >
        RATING
      </label>
      <input
        class="input"
        name="rating"
        placeholder="Rating here"
        type="number"
        value="0"
      />
    </div>
    <div
      class="buttons-area"
    >
      <button
        class="button default"
        data-testid="RESET"
      >
        RESET
      </button>
      <button
        class="button primary"
        data-testid="SUBMIT"
        type="submit"
      >
        SUBMIT
      </button>
    </div>
  </form>
</DocumentFragment>
`;
